pkg/router/crl/crl.go

```diff
@@ -0,0 +1,238 @@
+			// Creating or updating the configmap with incomplete data would compromise security by potentially
+			// permitting revoked certificates.
+			return nextCRLUpdate, fmt.Errorf("failed to get certificate revocation list for certificate key %s: %w", subjectKeyId, err)
+		} else if crl != nil {
+			crls[subjectKeyId] = crl
+			subjectKeyIds = append(subjectKeyIds, subjectKeyId)
+			log.Info("new certificate revocation list", "subject key identifier", subjectKeyId, "next update", crl.TBSCertList.NextUpdate.Format(time.RFC3339))
+			if (nextCRLUpdate.IsZero() || crl.TBSCertList.NextUpdate.Before(nextCRLUpdate)) && crl.TBSCertList.NextUpdate.After(now) {
+				nextCRLUpdate = crl.TBSCertList.NextUpdate
+			}
+			updated = true
+		}
+	}
+
+	if len(subjectKeyIds) == 0 {
+		if !crlFileExists {
+			// Make sure the crl file exists, even if it's empty
+			return nextCRLUpdate, os.WriteFile(crlFilename, []byte(""), crlFilePermissions)
+		} else {
+			return nextCRLUpdate, nil
+		}
+	}
+
+	// If any CRLs changed, encode the CRLs and write to `crlFilename`
+	if updated {
+		buf := &bytes.Buffer{}
+		for _, subjectKeyId := range subjectKeyIds {
+			asn1Data, err := asn1.Marshal(*crls[subjectKeyId])
+			if err != nil {
+				return nextCRLUpdate, fmt.Errorf("failed to encode ASN.1 for CRL for certificate key %s: %w", subjectKeyId, err)
+			}
+			block := &pem.Block{
+				Type:  "X509 CRL",
+				Bytes: asn1Data,
+			}
+			if err := pem.Encode(buf, block); err != nil {
+				return nextCRLUpdate, fmt.Errorf("failed to encode PEM for CRL for certificate key %s: %w", subjectKeyId, err)
+			}
+		}
+		crlData := buf.String()
+
+		return nextCRLUpdate, os.WriteFile(crlFilename, []byte(crlData), crlFilePermissions)
+	}
+
+	return nextCRLUpdate, nil
+}
+
+// getCRL gets a certificate revocation list using the provided distribution points and returns the certificate list.
+func getCRL(distributionPoints []string, now time.Time) (*pkix.CertificateList, error) {
+	var errs []error
+	for _, distributionPoint := range distributionPoints {
+		// The distribution point is typically a URL with the "http" scheme.  "https" is generally not used because the
+		// certificate list is signed, and because using TLS to get the certificate list could introduce a circular
+		// dependency (cannot use TLS without the revocation list, and cannot get the revocation list without using
+		// TLS).
+		//
+		// TODO Support ldap.
+		switch {
+		case strings.HasPrefix(distributionPoint, "http:"):
+			log.Info("retrieving CRL distribution point", "distribution point", distributionPoint)
+			crl, err := getHTTPCRL(distributionPoint)
+			if err != nil {
+				errs = append(errs, fmt.Errorf("error getting %q: %w", distributionPoint, err))
+				continue
+			}
+			if crl.TBSCertList.NextUpdate.Before(now) {
+				log.Info("crl expired. trying next distribution point", "nextUpdate", crl.TBSCertList.NextUpdate.Format(time.RFC3339))
+				continue
+			}
```

Comment:
New feature?  * grin *.  

If there isn't another distribution point to try, then this log message is misleading.  However, if all distribution points have expired data, then you should track this condition as an error.
```go
	for i, distributionPoint := range distributionPoints {
	// ...
```
```suggestion
			if crl.TBSCertList.NextUpdate.Before(now) {
				if i+1 < len(distributionPonts) {
					log.Info("distribution point has an expired CRL; trying next distribution point", "nextUpdate", crl.TBSCertList.NextUpdate.Format(time.RFC3339))
				} else {
					errs = append(errs, fmt.Errorf("all reachable distribution points publish expired CRLs"))
				}
				continue
			}
```
Alternatively, instead of returning an error here, the caller could treat a nil CRL value as an error condition.

Reply:
Yes, new feature :) It's not technically required to try other distribution points if one fails (assuming I understand RFC-5820 correctly), but I think this will help make this more robust. I probably should have called it out in the PR description, though.